### ISSUE TITLE
protoc --proto_out

### ISSUE THREAD

**Issue by jkopczyn** [protocolbuffers/protobuf#7409]
Projected use case:

I want to use a particular message type in several languages, on a machine with limited network access. Rather than compile the .proto into every language I'm going to use in advance, I would like to 'compile' the primary .proto file which contains that message definition into a new .proto which includes the primary file and also the definitions of all deps needed to compile it. I can then transmit this single file to the server which will use it, which can compile it to whichever languages it needs.

My specific use case is weird, narrow, and not quite according to the description above - I need to source a single file through a JS script (protobufjs.load) on a machine I can't access directly and which runs the script sandboxed (BigQuery JS UDFs) to prevent any outside network traffic.

**Comment by acozzette** [comment]
You can use `--descriptor_set_out` to parse the .proto files and save them all to a file in the form of a serialized `FileDescriptorSet`. Then on the isolated machine you can call `protoc` again with `--descriptor_set_in` and pass this file.

**Comment by jkopczyn** [comment]
Tried that already, actually. AFAICT it didn't work; it had a list of
source files it needed but not their contents.

**Comment by acozzette** [comment]
I think you may need to explicitly list all the dependency files when you call protoc with `--descriptor_set_out`.

**Comment by lukesneeringer** [comment]
As a note, I have wanted this several times, for various reasons.

1. We have a proto sanitizer (written in Java) that strips internal annotations and posts the sanitized proto to GitHub. Right now this is handled through templates.
2. I want to write a versioned proto manipulation pipeline that would do something similar, but for stripping things out of versions where they do not belong. Otherwise same principle.
3. I wrote a prototype for generating RPC services from resources that also wanted this.

Basically, there are legitimate reasons to want to input a proto, get the FileDescriptorSet, modify it, and write back the _actual proto files_.

**Comment by ruanspies** [comment]
@lukesneeringer - I agree 100%. Has anyone found any workarounds for this?

**Comment by deannagarcia** [comment]
As Adam said, I think `--descriptor_set_out` should be the solution to this. If you only saw the source files, then you probably need to pass all dependency files in your protoc call. If this still doesn't fix it, please reopen with information on how you're calling that and what output you get/how it differs from what you want.
